# Dataset Analysis

In [1]:
import anndata as ad
import scanpy as sc
from skopt.space import Real, Categorical
import pandas as pd

# Load training data
adata = ad.io.read_h5ad('../data/CellTypistDataset/global_annotated.h5ad')

# Filter blood data
adata = adata[adata.obs['Organ'] == 'BLD'].copy()
print(adata)


# Preprocessing

# mitochondrial genes, "MT-" for human, "Mt-" for mouse
adata.var["mt"] = adata.var_names.str.startswith("MT-")
# ribosomal genes
adata.var["ribo"] = adata.var_names.str.startswith(("RPS", "RPL"))
# hemoglobin genes
adata.var["hb"] = adata.var_names.str.contains("^HB[^(P)]")

sc.pp.calculate_qc_metrics(adata, qc_vars=["mt", "ribo", "hb"], inplace=True, log1p=True)

# Remove mitochondrial, ribosomal and hemoglobin
adata = adata[:, ~adata.var["mt"]].copy()
adata = adata[:, ~adata.var["ribo"]].copy()
adata = adata[:, ~adata.var["hb"]].copy()

# Doublet Detection
sc.pp.scrublet(adata, batch_key="Donor")


# Normalization

# Saving count data
adata.layers["counts"] = adata.X.copy()

# Normalizing to median total counts
sc.pp.normalize_total(adata)
# Logarithmize the data
sc.pp.log1p(adata)


# Filtering Highly variable genes
print('Before filtering highly variable genes ---')
print(adata)

sc.pp.highly_variable_genes(adata, n_top_genes=36473)

# Apply filter
adata = adata[:, adata.var['highly_variable']].copy()

print('After filtering highly variable genes ---')
print(adata)

# Create train test split

# All Donors: ['621B', '637C', 'A35', 'A36', 'D496', 'D503']
donor_train = ['637C', 'A35', 'A36', 'D503']
donor_test = ['621B', 'D496']

adata_train = adata[
    adata.obs["Donor"].isin(donor_train)
].copy()

adata_test = adata[
    adata.obs["Donor"].isin(donor_test)
].copy()

# Check split
print(adata_train.obs['Donor'].unique())
print(adata_test.obs['Donor'].unique())

# Prepare Data for training
X_train = adata_train.X#.to_df()
gene_names_train = adata_train.var_names
y_train = adata_train.obs['Manually_curated_celltype']

X_test = adata_test.X#.to_df()
gene_names_train = adata_train.var_names
y_test = adata_test.obs['Manually_curated_celltype']

AnnData object with n_obs × n_vars = 27620 × 36473
    obs: 'Organ', 'Donor', 'Chemistry', 'Cell_category', 'Predicted_labels_CellTypist', 'Majority_voting_CellTypist', 'Majority_voting_CellTypist_high', 'Manually_curated_celltype', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'n_genes', 'doublet_score', 'predicted_doublet', 'scumi-annotation'
    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'scrublet'
    obsm: 'X_umap'
Before filtering highly variable genes ---
AnnData object with n_obs × n_vars = 27620 × 36473
    obs: 'Organ', 'D

/tmp/ipykernel_7111/583758881.py:49: UserWarning: `n_top_genes` (=36473) > number of normalized dispersions (=27060), returning all genes with normalized dispersions.
  sc.pp.highly_variable_genes(adata, n_top_genes=36473)


After filtering highly variable genes ---
AnnData object with n_obs × n_vars = 27620 × 27060
    obs: 'Organ', 'Donor', 'Chemistry', 'Cell_category', 'Predicted_labels_CellTypist', 'Majority_voting_CellTypist', 'Majority_voting_CellTypist_high', 'Manually_curated_celltype', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'n_genes', 'doublet_score', 'predicted_doublet', 'scumi-annotation'
    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'
    uns: 'scrublet', 'log1p', 'hvg'
    obsm: 'X

In [2]:
classes = adata.obs['Manually_curated_celltype']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

Manually_curated_celltype
Tnaive/CM_CD4               6609
Classical monocytes         5850
Teffector/EM_CD4            4331
NK_CD16+                    3539
Tem/emra_CD8                1983
Tnaive/CM_CD8               1182
Tregs                        798
Trm/em_CD8                   635
NK_CD56bright_CD16-          445
T_CD4/CD8                    270
Nonclassical monocytes       236
MNP/T doublets               229
Cycling T&NK                 215
Naive B cells                194
Memory B cells               171
Megakaryocytes               163
Tfh                          111
MAIT                         107
DC2                          100
Plasma cells                  66
T/B doublets                  62
Progenitor                    61
Tgd_CRTAM+                    47
pDC                           43
Trm_Th1/Th17                  35
Plasmablasts                  35
Mast cells                    22
ABCs                          21
ILC3                          15
DC1              

Cells remaining after filtering min_samples=150
Tnaive/CM_CD4               6609
Classical monocytes         5850
Teffector/EM_CD4            4331
NK_CD16+                    3539
Tem/emra_CD8                1983
Tnaive/CM_CD8               1182
Tregs                        798
Trm/em_CD8                   635
NK_CD56bright_CD16-          445
T_CD4/CD8                    270
Nonclassical monocytes       236
MNP/T doublets               229
Cycling T&NK                 215
Naive B cells                194
Memory B cells               171
Megakaryocytes               163

In [3]:
classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

scumi-annotation
CD4+ T cell            13393
CD14+ monocyte          6481
Natural killer cell     4087
Cytotoxic T cell        2704
B cell                   395
Megakaryocyte            269
Dendritic cell           185
Plasma cell              106
Name: count, dtype: int64


Missing cells from marker genes:
CD16+ monocyte
Plasmacytoid dendritic cell

In [11]:
import anndata as ad
import scanpy as sc
from skopt.space import Real, Categorical
import pandas as pd

# Load training data
adata = ad.io.read_h5ad('../data/humancellatlas/5f29c29a-51c6-435c-8ff0-2b2a9d05ebee/BL_standard_design_annotated_30_000_samples_tmp_test.h5ad')

print(adata)


# Preprocessing

# mitochondrial genes, "MT-" for human, "Mt-" for mouse
adata.var["mt"] = adata.var_names.str.startswith("MT-")
# ribosomal genes
adata.var["ribo"] = adata.var_names.str.startswith(("RPS", "RPL"))
# hemoglobin genes
adata.var["hb"] = adata.var_names.str.contains("^HB[^(P)]")

sc.pp.calculate_qc_metrics(adata, qc_vars=["mt", "ribo", "hb"], inplace=True, log1p=True)

# Remove mitochondrial, ribosomal and hemoglobin
adata = adata[:, ~adata.var["mt"]].copy()
adata = adata[:, ~adata.var["ribo"]].copy()
adata = adata[:, ~adata.var["hb"]].copy()

# Doublet Detection
sc.pp.scrublet(adata, batch_key="Donor")


# Normalization

# Saving count data
adata.layers["counts"] = adata.X.copy()

# Normalizing to median total counts
sc.pp.normalize_total(adata)
# Logarithmize the data
sc.pp.log1p(adata)


# Filtering Highly variable genes
print('Before filtering highly variable genes ---')
print(adata)

sc.pp.highly_variable_genes(adata, n_top_genes=36473)

# Apply filter
adata = adata[:, adata.var['highly_variable']].copy()

print('After filtering highly variable genes ---')
print(adata)

AnnData object with n_obs × n_vars = 30000 × 25696
    obs: 'n_genes', 'Channel', 'n_counts', 'percent_mito', 'scale', 'Group', 'leiden_labels', 'Donor', 'doublet_score', 'pred_dbl', 'anno', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'predicted_doublet', 'scumi-annotation'
    var: 'featureid', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'Channels', 'Groups', 'PCs', 'W_pca_harmony', 'c2gid', 'df_qcplot', 'genome', 'gncells', 'leiden_resolution', 'modality', 'ncells', 'norm_count', 'pca', 'pca_features', 'pca_harmony_knn_distances', 'pc

/tmp/ipykernel_50810/3233603281.py:47: UserWarning: `n_top_genes` (=25696) > number of normalized dispersions (=21628), returning all genes with normalized dispersions.
  sc.pp.highly_variable_genes(adata, n_top_genes=36473)


After filtering highly variable genes ---
AnnData object with n_obs × n_vars = 30000 × 21628
    obs: 'n_genes', 'Channel', 'n_counts', 'percent_mito', 'scale', 'Group', 'leiden_labels', 'Donor', 'doublet_score', 'pred_dbl', 'anno', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'predicted_doublet', 'scumi-annotation'
    var: 'featureid', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'
    uns: 'Channels', 'Groups', 'PCs', 'W_pca_harmony', 'c2gid', 'df_qcplot', 'genome', 'gncells', 'leiden

In [12]:
classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

scumi-annotation
CD4+ T cell                    13677
Cytotoxic T cell                4843
B cell                          3960
CD14+ monocyte                  3135
Natural killer cell             2751
CD16+ monocyte                   908
Dendritic cell                   529
Plasmacytoid dendritic cell       88
Megakaryocyte                     86
Plasma cell                       23
Name: count, dtype: int64


In [10]:
classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

scumi-annotation
CD4+ T cell                    13677
Cytotoxic T cell                4843
B cell                          3960
CD14+ monocyte                  3135
Natural killer cell             2751
CD16+ monocyte                   908
Dendritic cell                   529
Plasmacytoid dendritic cell       88
Megakaryocyte                     86
Plasma cell                       23
Name: count, dtype: int64


In [8]:
classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

scumi-annotation
CD4+ T cell                    6912
Cytotoxic T cell               2393
B cell                         1963
CD14+ monocyte                 1582
Natural killer cell            1335
CD16+ monocyte                  464
Dendritic cell                  252
Plasmacytoid dendritic cell      55
Megakaryocyte                    44
Name: count, dtype: int64


In [18]:
import anndata as ad
import scanpy as sc
from skopt.space import Real, Categorical
import pandas as pd

# Load training data
adata_preprocessed = ad.io.read_h5ad('../data/human_immune_health_atlas/human_immune_health_atlas_full_annotated.h5ad')
adata = adata_preprocessed#.raw.to_adata()

# Filter blood data
#adata = adata[adata.obs['Organ'] == 'BLD'].copy()
print(adata)


# Preprocessing

# mitochondrial genes, "MT-" for human, "Mt-" for mouse
adata.var["mt"] = adata.var_names.str.startswith("MT-").to_numpy(dtype=bool)
# ribosomal genes
adata.var["ribo"] = adata.var_names.str.startswith(("RPS", "RPL")).to_numpy(dtype=bool)
# hemoglobin genes
adata.var["hb"] = adata.var_names.str.contains("^HB[^(P)]").to_numpy(dtype=bool)

sc.pp.calculate_qc_metrics(adata, qc_vars=["mt", "ribo", "hb"], inplace=True, log1p=True)

# Remove mitochondrial, ribosomal and hemoglobin
adata = adata[:, ~adata.var["mt"]].copy()
adata = adata[:, ~adata.var["ribo"]].copy()
adata = adata[:, ~adata.var["hb"]].copy()

# Doublet Detection
if "batch_id" not in adata.obs:
    print(
        "WARNUNG: 'batch_id' existiert nicht! Vorhandene Spalten:",
        adata.obs.columns.tolist(),
    )
# 1. Remove NaN values from batch_id after filtering
adata = adata[adata.obs["batch_id"].notna()].copy()

# 2. Filter empty cells and genes
sc.pp.filter_cells(adata, min_genes=1)
sc.pp.filter_genes(adata, min_cells=1)

# 3. Exclude too small batches
#batch_counts = adata.obs["batch_id"].value_counts()
#valid_batches = batch_counts[batch_counts >= 10].index
#adata = adata[adata.obs["batch_id"].isin(valid_batches)].copy()

# 3. Remove unused batch_ids
if str(adata.obs["batch_id"].dtype) == "category":
    adata.obs["batch_id"] = (
        adata.obs["batch_id"].cat.remove_unused_categories()
    )

# 4. Detect Doublets
#sc.pp.scrublet(adata, batch_key="Donor")
sc.pp.scrublet(adata, batch_key="batch_id")


# Normalization

# Saving count data
adata.layers["counts"] = adata.X.copy()

# Normalizing to median total counts
sc.pp.normalize_total(adata)
# Logarithmize the data
sc.pp.log1p(adata)


# Filtering Highly variable genes
print('Before filtering highly variable genes ---')
print(adata)

sc.pp.highly_variable_genes(adata, n_top_genes=36473)

# Apply filter
adata = adata[:, adata.var['highly_variable']].copy()

print('After filtering highly variable genes ---')
print(adata)

AnnData object with n_obs × n_vars = 31824 × 23921
    obs: 'cohort.cohortGuid', 'sample.sampleKitGuid', 'specimen.specimenGuid', 'pipeline.fileGuid', 'subject.subjectGuid', 'subject.biologicalSex', 'subject.birthYear', 'subject.ageAtFirstDraw', 'subject.ageGroup', 'subject.race', 'subject.ethnicity', 'subject.cmv', 'subject.bmi', 'sample.visitName', 'sample.drawYear', 'sample.subjectAgeAtDraw', 'batch_id', 'pool_id', 'chip_id', 'well_id', 'barcodes', 'original_barcodes', 'cell_name', 'n_reads', 'n_umis', 'n_genes', 'total_counts_mito', 'pct_counts_mito', 'doublet_score', 'AIFI_L1', 'AIFI_L2', 'AIFI_L3', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_co

In [ ]:
scumi-annotation
CD4+ T cell            13393
CD14+ monocyte          6481
Natural killer cell     4087
Cytotoxic T cell        2704
B cell                   395
Megakaryocyte            269
Dendritic cell           185
Plasma cell              106

In [20]:
classes = adata.obs['scumi-annotation']

classes_series = pd.Series(classes)

class_counts = classes_series.value_counts()
print(class_counts)

scumi-annotation
Natural killer cell    11772
Cytotoxic T cell       10385
B cell                  3865
CD4+ T cell             2513
CD16+ monocyte          1885
CD14+ monocyte          1323
Megakaryocyte             81
Name: count, dtype: int64


In [21]:
classes = adata.obs['AIFI_L1']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

AIFI_L1
T cell             18531
Monocyte            6034
B cell              3257
NK cell             2512
DC                  1355
Platelet              79
Progenitor cell       22
Erythrocyte           20
ILC                   14
Name: count, dtype: int64


In [22]:
classes = adata.obs['AIFI_L2']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

AIFI_L2
Memory CD4 T cell        6326
CD14 monocyte            5161
Naive CD4 T cell         4350
Memory CD8 T cell        3369
CD56dim NK cell          2306
Naive CD8 T cell         1756
Naive B cell             1472
MAIT                     1276
Plasma cell               962
cDC2                      777
gdT                       659
Treg                      643
CD16 monocyte             556
Memory B cell             515
pDC                       496
Intermediate monocyte     317
Transitional B cell       178
CD56bright NK cell        152
Effector B cell           130
Proliferating T cell       79
Platelet                   79
Proliferating NK cell      54
cDC1                       49
DN T cell                  40
CD8aa                      33
ASDC                       33
Progenitor cell            22
Erythrocyte                20
ILC                        14
Name: count, dtype: int64


In [17]:
classes = adata.obs['AIFI_L3']

classes_series = pd.Series(classes)

min_samples = 5
class_counts = classes_series.value_counts()
print(class_counts)

AIFI_L3
Core CD14 monocyte         4563
Core naive CD4 T cell      3963
CM CD4 T cell              3696
GZMK- CD56dim NK cell      1901
Core naive CD8 T cell      1707
                           ... 
CLP cell                      7
ISG+ naive CD8 T cell         6
Activated memory B cell       4
KLRF1- effector Vd1 gdT       3
BaEoMaP cell                  1
Name: count, Length: 71, dtype: int64
